File: sdk/php/tests/Unit/DSNParserTest.php
<?php

declare(strict_types=1);

namespace WorkerSQL\Tests\Unit;

use PHPUnit\Framework\TestCase;
use WorkerSQL\DSNParser;
use WorkerSQL\ValidationException;

class DSNParserTest extends TestCase
{
    public function testParseBasicDSN(): void
    {
        $dsn = 'workersql://api.workersql.com/mydb';
        $parsed = DSNParser::parse($dsn);

        $this->assertEquals('workersql', $parsed['protocol']);
        $this->assertEquals('api.workersql.com', $parsed['host']);
        $this->assertEquals('mydb', $parsed['database']);
        $this->assertNull($parsed['port']);
        $this->assertNull($parsed['username']);
        $this->assertNull($parsed['password']);
    }

    public function testParseDSNWithCredentials(): void
    {
        $dsn = 'workersql://user:pass@api.workersql.com/mydb';
        $parsed = DSNParser::parse($dsn);

        $this->assertEquals('user', $parsed['username']);
        $this->assertEquals('pass', $parsed['password']);
        $this->assertEquals('api.workersql.com', $parsed['host']);
        $this->assertEquals('mydb', $parsed['database']);
    }

    public function testParseDSNWithPort(): void
    {
        $dsn = 'workersql://api.workersql.com:8787/mydb';
        $parsed = DSNParser::parse($dsn);

        $this->assertEquals('api.workersql.com', $parsed['host']);
        $this->assertEquals(8787, $parsed['port']);
        $this->assertEquals('mydb', $parsed['database']);
    }

    public function testParseDSNWithQueryParameters(): void
    {
        $dsn = 'workersql://api.workersql.com/mydb?apiKey=abc123&ssl=false&timeout=5000';
        $parsed = DSNParser::parse($dsn);

        $this->assertEquals('abc123', $parsed['params']['apiKey']);
        $this->assertEquals('false', $parsed['params']['ssl']);
        $this->assertEquals('5000', $parsed['params']['timeout']);
    }

    public function testParseDSNWithSpecialCharacters(): void
    {
        $dsn = 'workersql://user%40name:p%40ss%3Aword@api.workersql.com/mydb';
        $parsed = DSNParser::parse($dsn);

        $this->assertEquals('user@name', $parsed['username']);
        $this->assertEquals('p@ss:word', $parsed['password']);
    }

    public function testParseDSNWithoutDatabase(): void
    {
        $dsn = 'workersql://api.workersql.com';
        $parsed = DSNParser::parse($dsn);

        $this->assertEquals('api.workersql.com', $parsed['host']);
        $this->assertNull($parsed['database']);
    }

    public function testInvalidProtocolThrowsException(): void
    {
        $this->expectException(ValidationException::class);
        $this->expectExceptionMessage('Invalid protocol');

        DSNParser::parse('mysql://api.workersql.com/mydb');
    }

    public function testEmptyDSNThrowsException(): void
    {
        $this->expectException(ValidationException::class);
        $this->expectExceptionMessage('non-empty string');

        DSNParser::parse('');
    }

    public function testGetApiEndpointHTTPSByDefault(): void
    {
        $parsed = DSNParser::parse('workersql://api.workersql.com/mydb');
        $endpoint = DSNParser::getApiEndpoint($parsed);

        $this->assertEquals('https://api.workersql.com/v1', $endpoint);
    }

    public function testGetApiEndpointHTTPWhenSSLFalse(): void
    {
        $parsed = DSNParser::parse('workersql://api.workersql.com/mydb?ssl=false');
        $endpoint = DSNParser::getApiEndpoint($parsed);

        $this->assertEquals('http://api.workersql.com/v1', $endpoint);
    }

    public function testGetApiEndpointIncludesPort(): void
    {
        $parsed = DSNParser::parse('workersql://api.workersql.com:8787/mydb');
        $endpoint = DSNParser::getApiEndpoint($parsed);

        $this->assertEquals('https://api.workersql.com:8787/v1', $endpoint);
    }

    public function testGetApiEndpointUsesProvidedEndpoint(): void
    {
        $parsed = DSNParser::parse(
            'workersql://api.workersql.com/mydb?apiEndpoint=https://custom.endpoint.com/api'
        );
        $endpoint = DSNParser::getApiEndpoint($parsed);

        $this->assertEquals('https://custom.endpoint.com/api', $endpoint);
    }
}

File: sdk/java/src/test/java/com/workersql/sdk/unit/DSNParserTest.java
package com.workersql.sdk.unit;

import com.workersql.sdk.util.DSNParser;
import com.workersql.sdk.util.ParsedDSN;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Unit tests for DSN Parser
 */
class DSNParserTest {

    @Test
    void testParseBasicDSN() {
        String dsn = "workersql://user:pass@api.workersql.com/mydb?apiKey=abc123";
        ParsedDSN parsed = DSNParser.parse(dsn);

        assertEquals("workersql", parsed.getProtocol());
        assertEquals("user", parsed.getUsername());
        assertEquals("pass", parsed.getPassword());
        assertEquals("api.workersql.com", parsed.getHost());
        assertNull(parsed.getPort());
        assertEquals("mydb", parsed.getDatabase());
        assertEquals("abc123", parsed.getParams().get("apiKey"));
    }

    @Test
    void testParseWithPort() {
        String dsn = "workersql://user:pass@localhost:8787/test?apiKey=key";
        ParsedDSN parsed = DSNParser.parse(dsn);

        assertEquals("localhost", parsed.getHost());
        assertEquals(8787, parsed.getPort());
        assertEquals("test", parsed.getDatabase());
    }

    @Test
    void testParseWithoutCredentials() {
        String dsn = "workersql://api.workersql.com/mydb?apiKey=key";
        ParsedDSN parsed = DSNParser.parse(dsn);

        assertNull(parsed.getUsername());
        assertNull(parsed.getPassword());
        assertEquals("api.workersql.com", parsed.getHost());
    }

    @Test
    void testParseWithMultipleParams() {
        String dsn = "workersql://host/db?apiKey=key&ssl=false&timeout=5000";
        ParsedDSN parsed = DSNParser.parse(dsn);

        assertEquals("key", parsed.getParams().get("apiKey"));
        assertEquals("false", parsed.getParams().get("ssl"));
        assertEquals("5000", parsed.getParams().get("timeout"));
    }

    @Test
    void testParseInvalidProtocol() {
        String dsn = "mysql://user:pass@host/db";
        assertThrows(IllegalArgumentException.class, () -> DSNParser.parse(dsn));
    }

    @Test
    void testParseInvalidFormat() {
        String dsn = "not-a-valid-dsn";
        assertThrows(IllegalArgumentException.class, () -> DSNParser.parse(dsn));
    }

    @Test
    void testParseEmptyDSN() {
        assertThrows(IllegalArgumentException.class, () -> DSNParser.parse(""));
        assertThrows(IllegalArgumentException.class, () -> DSNParser.parse(null));
    }

    @Test
    void testParseInvalidPort() {
        String dsn = "workersql://host:99999/db";
        assertThrows(IllegalArgumentException.class, () -> DSNParser.parse(dsn));
    }

    @Test
    void testStringify() {
        ParsedDSN parsed = new ParsedDSN(
            "workersql",
            "user",
            "pass",
            "api.workersql.com",
            443,
            "mydb",
            java.util.Collections.singletonMap("apiKey", "abc123")
        );

        String dsn = DSNParser.stringify(parsed);
        assertTrue(dsn.contains("workersql://"));
        assertTrue(dsn.contains("user:pass@"));
        assertTrue(dsn.contains("api.workersql.com:443"));
        assertTrue(dsn.contains("/mydb"));
        assertTrue(dsn.contains("apiKey=abc123"));
    }

    @Test
    void testGetApiEndpointWithSSL() {
        ParsedDSN parsed = new ParsedDSN(
            "workersql",
            null,
            null,
            "api.workersql.com",
            443,
            "mydb",
            java.util.Collections.emptyMap()
        );

        String endpoint = DSNParser.getApiEndpoint(parsed);
        assertEquals("https://api.workersql.com:443/v1", endpoint);
    }

    @Test
    void testGetApiEndpointWithoutSSL() {
        ParsedDSN parsed = new ParsedDSN(
            "workersql",
            null,
            null,
            "localhost",
            8787,
            "test",
            java.util.Collections.singletonMap("ssl", "false")
        );

        String endpoint = DSNParser.getApiEndpoint(parsed);
        assertEquals("http://localhost:8787/v1", endpoint);
    }

    @Test
    void testGetApiEndpointFromParams() {
        ParsedDSN parsed = new ParsedDSN(
            "workersql",
            null,
            null,
            "host",
            null,
            "db",
            java.util.Collections.singletonMap("apiEndpoint", "https://custom.endpoint.com/v2")
        );

        String endpoint = DSNParser.getApiEndpoint(parsed);
        assertEquals("https://custom.endpoint.com/v2", endpoint);
    }

    @Test
    void testURLEncoding() {
        String dsn = "workersql://user%40domain:p%40ss@host/db%20name?key=value%20with%20spaces";
        ParsedDSN parsed = DSNParser.parse(dsn);

        assertEquals("user@domain", parsed.getUsername());
        assertEquals("p@ss", parsed.getPassword());
        assertEquals("db name", parsed.getDatabase());
        assertEquals("value with spaces", parsed.getParams().get("key"));
    }
}
